### PR TITLE
feat(catalog-backend): Add catalog processing pipeline parameters

### PR DESCRIPTION
**Pull request by deejay1**
## Hey, I just made a Pull Request!

Allows for settings the low and high watermarks for the pipeline.

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by allegro-dbargowski — APPROVED** [review]

**Review by adowsky — APPROVED** [review]

**Review by awanlin — CHANGES_REQUESTED** [review]
Hi @deejay1, same with this one can you please add documentation for this change, here would be fine: https://github.com/backstage/backstage/blob/master/docs/features/software-catalog/configuration.md#processing-interval

**Review by vinzscam — APPROVED** [review]

**Comment by deejay1** [comment]
@awanlin could you rereview it ?

**Comment by deejay1** [comment]
@freben could you take a look? it would be great if this hit todays release

File: docs/features/software-catalog/configuration.md [hunk 1]
```diff
@@ -192,6 +192,24 @@ Setting this value too low risks exhausting rate limits on external systems that
...
+  Once the number of in-flight tasks reaches this count, more tasks will be loaded in.
+- `highWatermark` defines the maximum number of items to process in parallel.
+
+Keep in mind that the `lowWatermark` has to be less than `highWatermark`.
+
+```yaml title="app-config.yaml"
+catalog:
+  pipeline:
+    lowWatermark: 5
+    highWatermark: 10
```
Comment (freben) [inline docs/features/software-catalog/configuration.md:210]: Could we make this

```yaml
catalog:
  processing:
    concurrency: { min: 5, max: 10 }
```

instead?

The reason I'm asking is, I've been thinking that it'd be good to move ALL processing loop related things into a `processing` key, and this naming also expresses more clearly what is going on.

Later down the line, we may move `processingInterval` into `processing.interval` etc as well, and maybe have a `processing.enabled` too, but do not do that as part of this PR.

File: plugins/catalog-backend/src/processing/DefaultCatalogProcessingEngine.ts [hunk 2]
```diff
@@ -129,9 +129,13 @@ export class DefaultCatalogProcessingEngine {
   }
 
   private startPipeline(): () => void {
+    const lowWatermark =
+      this.config.getOptionalNumber('catalog.pipeline.lowWatermark') ?? 5;
+    const highWatermark =
+      this.config.getOptionalNumber('catalog.pipeline.highWatermark') ?? 10;
```
Comment (freben) [inline plugins/catalog-backend/src/processing/DefaultCatalogProcessingEngine.ts:135]: Should throw an exception if min isn't strictly smaller than max.

I'm also not sure, should you really be permitted to define just one of them? That'll just cause problems if we want to change the defaults - then people's installations may start to break if that violates this check. So, maybe `concurrency` should be optional, but not `min` and `max` inside it?

**Review by freben — COMMENTED** [review]
Were you adding this just for completeness, or were you really having issues from it not being configurable? We wanted to leave as few configurability parameters as possible in here.

File: docs/features/software-catalog/configuration.md (hunk 1 above)
Comment (deejay1) [reply docs/features/software-catalog/configuration.md]: done

**Comment by deejay1** [comment]
> Were you adding this just for completeness, or were you really having issues from it not being configurable? We wanted to leave as few configurability parameters as possible in here.

Unfortunately we really do need this and we're really not looking forward to forking the catalog due to the need to change two hardcoded parameters, which are a good tunable when working with thousands of entities. Right now every processing run tends to take over 15 minutes and tuning this should allow to reduce that and be more responsive for our users

**Comment by drodil** [comment]
@freben @vinzscam any chance this could make it to 1.32.0?

**Review by awanlin — APPROVED** [review]

**Comment by Rugvip** [comment]
@deejay1 what parameters work well for you if you tweak this with Yarn patches? We've picked the current values very arbitrarily and would happily update to whatever is optimal for your setup as that is probably going to benefit most deployments out there.

**Comment by drodil** [comment]
@Rugvip I think it depends on the catalog size and thus I am in favor of making this configurable. Sure the defaults should be optimal values for most use cases.

I think these values are one of those things you can only find out with try and error by running the instance for some time with each change.

**Comment by Rugvip** [comment]
I don't quite see how it could depend on catalog size tbh, we're really just looking for maximum concurrency and throughput right? If one instance grabs too many items that'd be bad, but if they're DB contention or idling that might be bad too.

I'm fairly confident that nobody will benefit from making this configurable in the long runt as it makes the overall setup more inflexible. I'm quite curious if there's a good indication that we could use to adjust these dynamically as well, for example based on how many items we actually manage to pick from the DB on average vs our throughput?

**Comment by deejay1** [comment]
Why does it make the setup more inflexible? For example we consider running only worker nodes of backstage to handle only tasks instead of everything, so we could tune them separately

**Pull request closed without merging**